### PR TITLE
Added a new 'snap' function to SimpleNumber that quantizes values

### PR DESCRIPTION
**Pull request by cianoc**
See:
https://github.com/supercollider/supercollider/pull/2826

Added a new snap function which 'quantizes' values. It's really just the existing quantize function, but which does the expected thing.

As part of this PR I added help for snap, added some clarification to quantize and also added a unit test for this new function.

I welcome discussion on the best name for this function, as well as what to do about the old 'quantize' function.

### CODE REVIEW

**Comment by telephon** [comment]
It would be good to add a comparison between `quantize` and `snap`, showing how both can be useful.

I could also imagine a `quantize` function that, if passed a negative quantum, would result in the `snap` behavior. This would be a typical way to formulate such symmetric distinctions.

**Comment by cianoc** [comment]
Well quantize will change any values bigger than (quantum+-tolerance) while snap will change any value smaller than (quantum+-tolerance). So I can certainly add that. There was talk of deprecating quantize just because it's horribly named (my proposed function 'snap' is what quantize should have been). Of course in an ideal world my new function would be 'quantize', and the current 'quantize' function would probably be called 'snap'. Backwards compatibility is annoying sometimes.

I'd also like to change the parameter name for quantize. I proposed 'grid' for snap, as it relates to the idea of a musical grid of beats (e.g. a sequencer/DAW). I really don't love that, but quantum is quite mathy and scary to non-math people (which is probably most users of SuperCollider, and certainly the ones most intimidated by it in the first place).

**Comment by brianlheim** [comment]
Quantize should be deprecated, for the reasons given in #2826:

- quantize does not do what most people will expect; in other words, its name is confusing
- what it does do is not a needed function for a vast majority of users
- it appears that quantize was incorrectly written, but needs to maintain backwards compatibility
- we should strongly encourage users to use a correctly behaving function rather than an incorrectly behaving function, and the best way to do that is through deprecation of the incorrect function

A possible alternative is to deprecate quantize and add a second function which does the same thing but is more appropriately named. However, I think you'll have a hard time coming up with a name for it, or even an example showing its utility. IMO there is none. I would rather just deprecate it.

> I proposed 'grid' for snap

I think grid is fine, although it isn't immediately intuitive to me. Technically speaking the value isn't a grid, it's the resolution of the grid; EnvelopeView has the same naming problem. I offer 'resolution' for consideration. A longer name is OK here since it's very unlikely someone would type out the parameter name when calling it - in other words, write `3.snap(resolution: 4)` instead of just `3.snap(4)`.

> I'd also like to change the parameter name for quantize.

Since quantize will be deprecated, changing the parameter name doesn't do much. I think it should be left as-is.

**Comment by telephon** [comment]
If we can imagine another way for a number to snap, the name probably should be `snapToGrid`. But otherwise I'm fine with `snap`. I hope nobody complains that a float snapped at him.

**Comment by muellmusik** [comment]
These are in some sense variants of round, yes? And possibly useful for more than the normal quantising stuff? Maybe they should be named roundSomething?

snapToGrid is nice, but grids need not be regular. A real snapToGrid might be useful, but would require a Grid object capable of representing more than equal steps.

**Comment by brianlheim** [comment]
I don't have a strong opinion on the name, but let's try to suggest concrete alternatives so the discussion doesn't derail.

**Comment by muellmusik** [comment]
roundWithTolerance?

**Comment by telephon** [comment]
we have `roundUp` and `roundDown`. Maybe something analog? Like `roundToward`? 

On the other hand, `snap` isn't a bad name for what it does. 

I can imagine many other rounding to grid functions, but they can almost always better be written directly as a function.

**Comment by patrickdupuis** [comment]
`roundAbout` ?

... sorry couldn't help myself

**Comment by telephon** [comment]
(I also needed to hold myself back!)

**Comment by jamshark70** [comment]
BTW, I never published this extension, but I use it *all the time* for FM modulator ratios:

```
SnapControlSpec : ControlSpec {
	var <>snap = 0.05;
	map { arg value;
		// maps a value from [0..1] to spec range
		var out = warp.map(value.clip(0.0, 1.0)),
		rounded = out.round(step);
		if((out absdif: rounded) <= snap) { ^rounded } { ^out }
	}
	unmap { arg value;
		// maps a value from spec range to [0..1]
		var rounded = value.round(step);
		if((value absdif: rounded) <= snap) { value = rounded };
		^warp.unmap(value.clip(clipLo, clipHi));
	}
	constrain { arg value;
		var rounded;
		value = value.asFloat.clip(clipLo, clipHi);
		rounded = value.round(step);
		if((value absdif: rounded) <= snap) { ^rounded } { ^value }
	}
}
```

`SnapControlSpec(0.5, 20, \lin, 0.5).snap_(0.05)`, e.g.

Or a panner with an indent in the center: `SnapControlSpec(-1, 1, \lin, 1).snap_(0.05)`

**Comment by cianoc** [comment]
> I offer 'resolution' for consideration. A longer name is OK here since it's very unlikely someone would type out the parameter name when calling it - in other words, write 3.snap(resolution: 4) instead of just 3.snap(4).

This seems like a no-brainer. Resolution it is. Nice one.

**Comment by cianoc** [comment]
roundToward isn't bad. roundWithTolerance would be perfect if it wasn't so long. gradualRound? There are two hard problems in Computer Science - naming things and cache invalidation.

I'd be inclined if we went down that route to replicate the old quantize function as snap (with better named parameters and help).

**Comment by jamshark70** [comment]
I won't hold my breath for graphics programs to start to `roundToward` grid lines instead of snapping to them...

**Comment by brianlheim** [comment]
my thought after seeing these suggestions is `softRound`

**Comment by cianoc** [comment]
I like softRound.

How about I create a copy of quantize called 'snap' which has the behavior that James described (snap to grid lines). And my function is softRound.  And I'll add better documentation for 'snap' than the current quantize function.

And quantize is deprecated.

**Comment by telephon** [comment]
`roundSoftly` would be found easier in autocomplete.

**Comment by brianlheim** [comment]
naming shouldn't be justified by current autocomplete functionality imho. i'd rather have code that makes sense and fits with naming conventions.

**Comment by cianoc** [comment]
Well there's currently a round and a roundUp (though not a roundDown - what's 'up' with that). So I guess it would fit into an already existing scheme if it was roundSoftly.

**Comment by telephon** [comment]
yes, that's what I actually wanted to say. But I'm fine with most variants.

**Comment by cianoc** [comment]
So do we think we can achieve some kind of consensus on naming?

I'd like to call it roundSoft, or roundSoftly, just because it would fit into the already existing naming convention for round. And use snap for 'snap' in the sense that James referenced it above. But I'm open to persuasion.

And I'll start a new pull request to add roundDown... Seriously, why's that not in there?

**Comment by brianlheim** [comment]
if you're looking for a convention, the other things in SC that have "soft" in their names are:

- PV_SoftWipe
- softPut
- softSet
- softVol_
- softclip

similarly, there are

- hardFreeAll
- hardRun
- hardStop

**Comment by cianoc** [comment]
So you had to bring facts to this discussion @brianlheim...

Okay I'll rename the function to roundSoftly, move the help up so that it's in the round section. And I'll create a copy of quantize called snap. And I'll update help appropriately.

Is that cool?

**Comment by cianoc** [comment]
new function name softRound, not roundSoftly. Dammit.

File: .gitmodules [hunk 1]
```diff
@@ -17,3 +17,6 @@
 [submodule "external_libraries/yaml-cpp"]
 	path = external_libraries/yaml-cpp
 	url = https://github.com/supercollider/yaml-cpp.git
+[submodule "external_libraries/link"]
```
Comment (telephon) [inline .gitmodules]: this shouldn't be part of this PR … Happens all the time! You'll need a submodule reset or similar.

**Comment by cianoc** [comment]
Sorry, messed up the merge. Should be fixed now.

**Comment by cianoc** [comment]
@telephon - sorry I screwed up the merge. My latest version has removed the reference to Ableton link which shouldn't have been in there, and I'm still not entirely sure how it got on that branch.

**Comment by jamshark70** [comment]
Hm, now I'm reading the help file, and `softRound` does what I thought `snap` would do...???

And I'm really sorry to spoil consensus, but I very strongly dislike the name `softRound`. It means nothing to me... whereas, for instance, studio musicians are very familiar with panpots that "snap" to center when the knob gets close enough to the center position, or volume faders that snap to unity gain.

IMO we've gotten these precisely backward. Users will be very surprised when they try to snap numbers to a grid and find that it doesn't apply to numbers close to the grid.

Help: What does "Values that are +-tolerance away from a strong::resolution:: value" mean? *Within* the tolerance, or further away ("at least")?

I'm afraid I might have muddied the waters by introducing the analogy to graphics programs. As far as I remember, most of them do simple rounding for "snap to grid." I would like it if they would "snap to grid" only when the pointer gets close enough to the grid, but I haven't seen that behavior. I'm sure it would be very unusual for graphics programs to snap to grid when you are far away from the grid, but allow you to be sloppy close to the grid. So "Conceptually this is the equivalent of 'snap' in a graphics program," in hindsight, I think is wrong.

Sorry, I can't agree to this PR as it is.

**Comment by cianoc** [comment]
@jamshark70 - clearly I did a bad job of writing the help for these as snap does the inverse of what you describe. Values close to the grid will be rounded. Values greater than tolerance will be left as they are. I can't speak for all graphics programs but I've certainly used graphics programs that offer this functionality.

I've made a few changes that might help clarify - otherwise I probably need to leave it for a few days so I can get some distance from this.

I am neutral on the names. Well that's not true. The perfect name for softlyRound would be quantize, but sadly that ship has sailed.

File: testsuite/classlibrary/TestSimpleNumber.sc [hunk 2]
```diff
@@ -0,0 +1,20 @@
+
+TestSimpleNumber : UnitTest {
+
+	test_snap {
+		var val = [ 0, 0.2, 0.4, 0.6, 0.8, 1, 1.2, 1.4, 1.6, 1.8, 2 ];
```
Comment (brianlheim) [inline testsuite/classlibrary/TestSimpleNumber.sc]: just a note for future tests (although if you don't mind changing this one too, i'd appreciate it): you don't need this many test values to get full coverage. three or four should be enough — 0, a multiple of resolution, a value within tolerance, and a value outside tolerance would be a good set.

you're also missing test cases for when resolution is 0 or negative, tolerance is negative, and strength is 0 or negative. even if these are not intended use cases, it's good to pin them down so that if someone rewrites the method in the future, we will know what corner cases may have changed and can use that information responsibly.

i'd be happy to write those tests (testing is a very calming activity for me).

also, this file conflicts now with the TestSimpleNumber I added in my `asTimeString` PR. i think you should be fine just rebasing this on 3.9 (`git rebase 3.9 topic/quantize-function`) and then force-pushing.

File: SCClassLibrary/Common/Math/SimpleNumber.sc [hunk 3]
```diff
@@ -160,6 +160,26 @@ SimpleNumber : Number {
 		}
 	}
 
+	snap { arg resolution = 1.0, tolerance = 0.05, strength = 1.0;
+		var round = round(this, resolution);
+		var diff = round - this;
+		if (abs(diff) < tolerance) {
+			^this + (strength * diff)
+		}{
```
Comment (brianlheim) [inline SCClassLibrary/Common/Math/SimpleNumber.sc]: i know this is copy-pasted, but please add a space between brackets here and in line 179 as well.

Comment (brianlheim) [reply SCClassLibrary/Common/Math/SimpleNumber.sc]: you could change the original as well if you want

File: SCClassLibrary/Common/Math/SimpleNumber.sc [hunk 4]
```diff
@@ -160,6 +160,26 @@ SimpleNumber : Number {
 		}
 	}
 
+	snap { arg resolution = 1.0, tolerance = 0.05, strength = 1.0;
```
Comment (brianlheim) [inline SCClassLibrary/Common/Math/SimpleNumber.sc]: i would question the name "tolerance" for the second parameter. usually, that indicates a range of untransformed values, but here it's actually describing the range of effect. i searched around for a name for this, and all i could come up with was "margin". what do you think?

File: HelpSource/Classes/SimpleNumber.schelp [hunk 5]
```diff
@@ -67,8 +67,61 @@ Greatest common divisor
 
 method:: round
 Round to multiple of aNumber
+
 method:: roundUp
-round up to a multiply of aNumber
+Round up to a multiply of aNumber. For roundDown use Link::Classes/SimpleNumber#-trunc#trunc::.
```
Comment (brianlheim) [inline HelpSource/Classes/SimpleNumber.schelp]: could you correct this to 'multiple' please?

File: HelpSource/Classes/SimpleNumber.schelp [hunk 6]
```diff
@@ -67,8 +67,61 @@ Greatest common divisor
...
+method:: snap
+Rounds the values strong::tolerance:: distance from resolution to a multiple of resolution. By using strong::tolerance:: and strong::strength:: you can control when values will be rounded, and by how much.
+
+Conceptually this is the equivalent of 'snap' in a graphics program. Values within a certain distance (strong::tolerance::) from a grid line are snapped to it. All other values are unchanged.
+
+argument::resolution
+Snap this value to a multiple of resolution. E.g. if you chose 1, then all values would be rounded to the nearest integer.
+
+argument::tolerance
+Only values that are greater +-strong::tolerance:: from a multiple of strong::resolution:: value will be changed. Values that are less than strong::tolerance:: will be unchanged.
```
Comment (brianlheim) [inline HelpSource/Classes/SimpleNumber.schelp]: fyi, documentation is UTF-8 encoded, so you could use ± if you wanted to.

**Comment by jamshark70** [comment]
@cianoc Ah, OK, thanks. Then, I'm OK with it.

Indeed, documentation is hard to write. Professional software companies have a separate documentation team, specifically because coders often don't think of what makes good documentation.  One side effect of not having a dedicated documentation team is that, for a lot of method documentation, you have to run the code example to understand what the help text is trying to say.

Anyway, it is clearer now by example.

File: SCClassLibrary/Common/Math/SimpleNumber.sc (hunk 4 above)
Comment (cianoc) [reply SCClassLibrary/Common/Math/SimpleNumber.sc]: I like margin as it provides a good visualization of what's going on. I shall change.

File: HelpSource/Classes/SimpleNumber.schelp (hunk 6 above)
Comment (cianoc) [reply HelpSource/Classes/SimpleNumber.schelp]: Ha! Done.

**Comment by cianoc** [comment]
I still need to rewrite the tests. Everything else should be done.

File: .gitmodules (hunk 1 above)
Comment (brianlheim) [reply .gitmodules]: @cianoc this still needs to be removed

File: SCClassLibrary/Common/Math/SimpleNumber.sc [hunk 7]
```diff
@@ -155,12 +155,32 @@ SimpleNumber : Number {
...
-		}{
+		} {
+			^this
+		}
+	}
+
+	snap { arg resolution = 1.0, margin = 0.05, strength = 1.0;
+		var round = round(this, resolution);
+		var diff = round - this;
+		if ( abs(diff) < margin ) {
```
Comment (brianlheim) [inline SCClassLibrary/Common/Math/SimpleNumber.sc]: why are there suddenly spaces in the parens now? also FWIW i think our semiformal style is to not add a space after `if` since it is also a method and not a true control structure as in many other languages:

`if(abs(diff) < margin)`

File: testsuite/classlibrary/TestSimpleNumber.sc [hunk 8]
```diff
@@ -102,4 +102,19 @@ TestSimpleNumber : UnitTest {
...
+		this.assertEquals(testF.(0.5,0,1), [ 0, 0, 0.5, 0.5, 1, 1, 1, 1.5, 1.5, 2, 2 ], "Snap test 2 failed.");
+
+		this.assertEquals(testF.(1,0.3,1), [ 0, 0.2, 0, 1, 0.8, 1, 1.2, 1, 2, 1.8, 2 ], "Snap test 3 failed.");
+
+		this.assertEquals(testF.(1,0,0.5), [ 0, 0.1, 0.2, 0.8, 0.9, 1, 1.1, 1.2, 1.8, 1.9, 2 ], "Snap test 4 failed.");
+
+		this.assertEquals(testF.(1,0.2,0.5), [ 0, 0.2, 0.2, 0.8, 0.8, 1, 1.2, 1.2, 1.8, 1.8, 2 ], "Snap test 5 failed.");
+
+	}
+}
```
Comment (brianlheim) [inline testsuite/classlibrary/TestSimpleNumber.sc]: missing an EOL here. we should have a low-priority issue that ensures the IDE always inserts a \n after the last line in the editor, we have way too many faulty commits due to this.

File: SCClassLibrary/Common/Math/SimpleNumber.sc (hunk 7 above)
Comment (cianoc) [reply SCClassLibrary/Common/Math/SimpleNumber.sc]: I have no idea how those spaces got there. Wasn't deliberate.

**Comment by cianoc** [comment]
I think this should address all the issues.

The data now covers all the edge cases (plus a little extra for my OCD paranoia), including negative values. Also have tests for values that should never be entered.

Apparently I don't find user tests soothing :-/

File: testsuite/classlibrary/TestSimpleNumber.sc [hunk 9]
```diff
@@ -102,4 +102,69 @@ TestSimpleNumber : UnitTest {
...
+
+		val = [-1, -0.5, 0, 1, 0.5, 1];
+		this.assertEquals(testF.(val,1,0,-0.5), [-1, -0.75, 0, 1, 0.25, 1], "softRound test 8 failed.");
+
+		this.assertEquals(testF.(val,1,0,0), val, "softRound test 10 failed.");
+		this.assertEquals(testF.(val,1,2,1), val, "softRound test 11 failed.");
+
+	}
+
+	test_Snap {
```
Comment (brianlheim) [inline testsuite/classlibrary/TestSimpleNumber.sc]: this should be lowercase "snap"

File: testsuite/classlibrary/TestSimpleNumber.sc [hunk 10]
```diff
@@ -102,4 +102,69 @@ TestSimpleNumber : UnitTest {
 		this.assert(actual == expected, "expected %, got %".format(expected, actual));
 	}
 
+	test_softRound {
+		var val;
+		var testF = {|vals, g, t, s| vals.collect({|num| num.softRound(g, t, s)})};
+
+		val = [-1.3, -1, -0.7, -0.5, -0.3, 0, 0.2, 0.5, 0.7, 1, 1.2, 1.8, 2];
+		this.assertEquals(testF.(val, 1,0,1), [ -1, -1, -1, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2 ], "softRound test 1 failed.");
```
Comment (brianlheim) [inline testsuite/classlibrary/TestSimpleNumber.sc]: keep in mind that `message` is posted even on a successful verbose run. The message should describe the test; these currently read like error messages

File: testsuite/classlibrary/TestSimpleNumber.sc [hunk 11]
```diff
@@ -102,4 +102,69 @@ TestSimpleNumber : UnitTest {
...
+
+		val = [-0.8, -0.7, -0.6 -0.4, -0.3, -0.2, 0, 0.2, 0.3, 0.4, 0.6, 0.7, 0.8, 1];
+		this.assertEquals(testF.(val,1,0.3,1), [ -0.8, -1, -1, -0.3, -0.2, 0, 0.2, 0.3, 0, 1, 1, 0.8, 1 ], "softRound test 3 failed.");
+
+		val = [-1, -0.8, -0.2, 0, 0.2, 0.8, 1];
+		this.assertEquals(testF.(val,1,0,0.5), [ -1, -0.9, -0.1, 0, 0.1, 0.9, 1 ], "softRound test 4 failed.");
+
+		// Testing weird edge cases that are probably not typical use cases for backwards compatability purposes.
+		val = [-1, -0.5, 0, 1, 0.5, 1];
+		this.assertEquals(testF.(val,0,0,1), val, "softRound test 5 failed."); // resolution value of 0 leave the original array unchanged.
```
Comment (brianlheim) [inline testsuite/classlibrary/TestSimpleNumber.sc]: could you please put these end of line comments on the line above instead?

File: testsuite/classlibrary/TestSimpleNumber.sc (hunk 10 above)
Comment (brianlheim) [reply testsuite/classlibrary/TestSimpleNumber.sc]: and again, testing way too many values IMO, it's very hard to tell what's going on in this method

Comment (cianoc) [reply testsuite/classlibrary/TestSimpleNumber.sc]: Ah, sorry used to unit test libraries that only show the message when it fails. I will modify. I see why people complain about the verbosity of the messages.

Each of the values is testing a very specific edge case where a bug (imo) could creep in. I could split these onto multiple lines, but I'm hesitant to reduce coverage.

Comment (brianlheim) [reply testsuite/classlibrary/TestSimpleNumber.sc]: thank you!

**Comment by brianlheim** [comment]
as an aside, this has definitely given me some good ideas for what to include in a unit testing guide. i'm sorry that you didn't have anything to go off of for this PR, that shouldn't happen in the future.

**Comment by cianoc** [comment]
That submodule keeps creeping back in. This should be fixed now.

**Comment by brianlheim** [comment]
NB - I realized at the last moment that this PR hadn't actually deprecated quantize yet, so I went ahead and pushed that change to your branch, @cianoc
